
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../microbiome/">
      
      
        <link rel="next" href="../ont/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.12">
    
    
      
        <title>ONT Assembly - Omics course</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.2afb09e1.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#ont-assembly" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Omics course" class="md-header__button md-logo" aria-label="Omics course" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Omics course
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              ONT Assembly
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Omics course" class="md-nav__button md-logo" aria-label="Omics course" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    Omics course
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Pathogen genomics
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../microbiome/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Microbiomes Practical
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    ONT Assembly
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    ONT Assembly
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      Overview
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#software-and-data-requirements" class="md-nav__link">
    <span class="md-ellipsis">
      Software and Data Requirements
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#read-qc" class="md-nav__link">
    <span class="md-ellipsis">
      Read QC
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#assembly" class="md-nav__link">
    <span class="md-ellipsis">
      Assembly
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#assembly-qc" class="md-nav__link">
    <span class="md-ellipsis">
      Assembly QC
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#annotation" class="md-nav__link">
    <span class="md-ellipsis">
      Annotation
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../ont/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Third Generation Sequencing
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../tb-genomics/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    TB Genomics
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      Overview
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#software-and-data-requirements" class="md-nav__link">
    <span class="md-ellipsis">
      Software and Data Requirements
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#read-qc" class="md-nav__link">
    <span class="md-ellipsis">
      Read QC
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#assembly" class="md-nav__link">
    <span class="md-ellipsis">
      Assembly
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#assembly-qc" class="md-nav__link">
    <span class="md-ellipsis">
      Assembly QC
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#annotation" class="md-nav__link">
    <span class="md-ellipsis">
      Annotation
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="ont-assembly">ONT Assembly</h1>
<h2 id="overview">Overview</h2>
<p>In this session we will be performing a de novo assembly of a genome using ONT reads. We will be using the <code>flye</code> assembler, which is specifically designed for long-read sequencing data. You will contrast these methods and the output of the assembly with the output of a short-read assembly using <code>spades</code> that you generated earlier in the course.</p>
<h2 id="software-and-data-requirements">Software and Data Requirements</h2>
<p>We will be using the following software and data in this activity:</p>
<ol>
<li>Nanoplot - to perform read quality control</li>
<li>Flye - to perform de novo assembly of the genome</li>
<li>Quast - to perform assembly quality control</li>
<li>Prokka - to annotate the assembly</li>
</ol>
<p>We will now use mamba to create an environment to install the required software. </p>
<div class="highlight"><pre><span></span><code>mamba create -n ont-assembly flye quast prokka nanoplot
</code></pre></div>
<p>After it has finished you can activate the environment using:</p>
<div class="highlight"><pre><span></span><code>mamba activate ont-assembly
</code></pre></div>
<p>We will also be using data from the <a href="https://journals.plos.org/globalpublichealth/article?id=10.1371/journal.pgph.0004099#sec013">this study</a>. </p>
<p>We are going to be assembling the data for one of the samples from this study. You should download the <abbr title="A format used to store sequence data with qualite values for each nucleotide">fastq</abbr> data from the ENA browser. The data is available under the accession <code>ERR12245902</code>. </p>
<p>Create a folder called <code>ont-assembly</code> in your <code>data</code> directory. After you download the data, you can move it from your <code>Downloads</code> folder to the <code>ont-assembly</code> folder. </p>
<h2 id="read-qc">Read QC</h2>
<p>Try to use nanoplot to summarise the quality of the reads. </p>
<div class="admonition question">
<p class="admonition-title">Question</p>
<p>What is the average read length of the reads?</p>
</div>
<h2 id="assembly">Assembly</h2>
<p>We will be using the <code>flye</code> assembler to perform the assembly. The a template of the command is as follows:</p>
<div class="highlight"><pre><span></span><code>flye --nano-hq &lt;fastq&gt; --out-dir &lt;output_dir&gt; --threads 2 
</code></pre></div>
<p>Where <code>&lt;fastq&gt;</code> is the path to the <abbr title="A format used to store sequence data with qualite values for each nucleotide">fastq</abbr> file, <code>&lt;output_dir&gt;</code> is the path to the output directory and <code>--threads</code> is the number of threads to use. Flye will create a file called <code>assembly.fasta</code> in the output directory. This file contains the assembled genome, which is a collection of contigs.</p>
<h2 id="assembly-qc">Assembly QC</h2>
<p>We will be using <code>quast</code> to perform the assembly quality control. The command is as follows:</p>
<div class="highlight"><pre><span></span><code>quast.py &lt;assembly.fasta&gt; -o &lt;output_dir&gt; --threads 2
</code></pre></div>
<p>Where <code>&lt;assembly.fasta&gt;</code> is the path to the assembly file and <code>&lt;output_dir&gt;</code> is the path to the output directory.</p>
<p>Quast will generate a number of files in the output directory. Have a look at the <code>report.html</code> file by opening it via the file browser or using the command:</p>
<div class="highlight"><pre><span></span><code>firefox &lt;output_dir&gt;/report.html
</code></pre></div>
<div class="admonition question">
<p class="admonition-title">Question</p>
<p>What is the N50 of the assembly?</p>
<p>What is the number of contigs in the assembly?</p>
<p>What is the total length of the assembly? How does this compare with the expected Mtb genome size?</p>
<p>How to these metrics compare to the short-read assembly you generated earlier in the course?</p>
</div>
<h2 id="annotation">Annotation</h2>
<p>We will be using <code>prokka</code> to annotate the assembly. The command is as follows:</p>
<div class="highlight"><pre><span></span><code>prokka --outdir &lt;output_dir&gt;  &lt;assembly.fasta&gt; --cpus 2
</code></pre></div>
<p>Where <code>&lt;output_dir&gt;</code> is the path to the output directory, <code>&lt;assembly.fasta&gt;</code> is the path to the assembly file and <code>--cpus</code> is the number of threads to use. In the output directory there will be several files, including files that end in:</p>
<ol>
<li>.fna - a <abbr title="A format used to store sequence data">fasta</abbr> file with the nucleotide sequence of the genome</li>
<li>.ffn - a <abbr title="A format used to store sequence data">fasta</abbr> file with the nucleotide sequence of the coding regions</li>
<li>.faa - a <abbr title="A format used to store sequence data">fasta</abbr> file with the amino acid sequence of the coding regions</li>
<li>.gff - a text file with the location of genes </li>
</ol>
<h1 id="comparing-protein-sequences">Comparing protein sequences</h1>
<p>Open up the <code>.faa</code> file in a text editor (by double clicking on it in the file browser or running <code>open &lt;filename&gt;</code> on the terminal). You will see that the file contains the protein sequences of the coding regions. </p>
<p>As we saw in the previous session, RpoB is a protein that is the target of the rifampicin antibiotic. We can use the protein sequence of RpoB to identify the presence of mutations that confer resistance to rifampicin. Look for the protein sequence of RpoB in the <code>.faa</code> file. You can search for it by using the command <code>ctrl + f</code> and typing <code>RNA polymerase subunit beta</code>. </p>
<p>On the terminal run <code>aliview</code> to open the aliview program. Copy in the protein sequence of RpoB from the <code>.faa</code> file. You can do this by selecting the sequence in the text editor and using <code>ctrl + c</code> to copy it. Then in aliview, use <code>ctrl + v</code> to paste it.</p>
<p>Now copy the protein sequence from https://mycobrowser.epfl.ch/genes/Rv0667 and paste it into aliview.</p>
<div class="admonition question">
<p class="admonition-title">Question</p>
<p>The two sequences are exactly same length, so in this case we don't need to align them. What are the differences between the two sequences? Are any of the differences in the region that is known to confer resistance to rifampicin? (use the table from the TB genomics session to help you with this).</p>
</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": ["content.tooltips", "navigation.instant"], "search": "../assets/javascripts/workers/search.f8cc74c7.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.c8b220af.min.js"></script>
      
    
  </body>
</html>